### DISCUSSION TITLE
Dual dGPU-Setup: one for Games, one just for encoding

### DISCUSSION THREAD

**Discussion by tibia1426** [ClassicOldSong/Apollo#155] [Q&A] [answered]
Hi!

I am currently testing Apollo with Moonlight in a scenario where I have a powerful GPU (AMD Radeon RX 7900 XT) dedicated solely to 3D rendering, with physical monitors connected to it, and a secondary GPU (Intel Arc A380) intended only for encoding for Apollo/Sunshine. Unfortunately, I have not been successful so far, or if it did work as desired, it might have been due to luck. 

I have already tested with headless mode on and off to duplicate the displays in Windows or to deactivate the physical monitors. Of course, I also tried specifying the GPU explicitly in Apollo. When I specify the GPU as the display adapter, I always get an error message after restarting Apollo that the GPU or encoder could not be found. Apollo/Sunshine can still use the GPU for encoding when only the virtual display is active and the physical monitors are not active. However, then I have the problem that games are also rendered using this GPU (A380). I then set in Windows graphics settings that only Apollo/Sunshine should use the A380. Indeed, the 7900 XT somehow calculates the game again - but apparently, the data still has to go through the "bottleneck" of the A380, presumably because it is responsible for the virtual display. 

Is there any way to achieve what I am trying to do are there any ideas? Or is it inherent in Apollo/Sunshine's design that display output must also occur through the GPU, and thus always has to pass through the bottleneck of the smaller A380 GPU (which is only running in 4x PCIe 2.0 mode) when it does the encoding? 

Thank you in advance! And thanks for all your hard work and dedication. I have been able to solve several problems with Apollo compared to Sunshine and find it very promising. Keep up the great work!

**Comment by kayti** [comment, marked as answer]
EDIT: My old anwser is already outdated. My current experience:
https://github.com/ClassicOldSong/Apollo/discussions/155#discussioncomment-13573097


OLD ANWSER:
Tried the same thing with RTX 3080 and ARC 380. In windows settings, somewhere under display there is GPU section where You can choose which GPU is used for applications. So normally stream using ARC, but set there RX as render. 

But in my case performance was subpar.

Uncompressed frame needs to be copied from RX to ARC (additional delay), in higher resolutions/framerate, uncompressed bandwitch hits PCIE 4 8x throughput. A380 has 8x lines, in my tests I was able to get both card into 8x/8x mode, but then RTX had slightly lower performance, and it was still too slow for A380 to deliver more than 4k@100 frames at best.

**Reply by tibia1426** [reply]
Okay, thank you for your response. I think that’s exactly my problem. I’m familiar with the setting in the Windows settings and have already tested it. It is also the case that obviously the AMD GPU performs the 3D rendering, but the performance is still poor because everything has to go through the bottleneck of the A380, as you described. Thank you for the deeper technical explanation. It makes sense. And since my Arc A380 is only running in PCIe 2.0 x4 mode due to the motherboard and chipset, the performance is, of course, unusable. I’m afraid I’ll have to abandon this attempt and return the A380.

**Reply by mreilaender** [reply]
@kayti I assume the limiting factor was the CPU or the motherboard since PCIE4 8x is theoretically capable of ~15 GB/s [Wikipedia] and 4k@120Hz HDR would need around 32.27 Gbps [[trychen.com](https://trychen.com/feature/video-bandwidth)]. Assuming raw video data needs to be transferred to (from the render GPU to the encoding GPU) and from (to the network interface for transmission) the encoding GPU PCIE x8 can't be the limiting factor here.

What's the model of your CPU/Motherboard that you used for testing?

**Reply by tibia1426** [reply]
> @kayti I assume the limiting factor was the CPU or the motherboard since PCIE4 8x is theoretically capable of ~15 GB/s [Wikipedia] and 4k@120Hz HDR would need around 32.27 Gbps [[trychen.com](https://trychen.com/feature/video-bandwidth)]. Assuming raw video data needs to be transferred to (from the render GPU to the encoding GPU) and from (to the network interface for transmission) the encoding GPU PCIE x8 can't be the limiting factor here.
> 
> What's the model of your CPU/Motherboard that you used for testing?

Asus Prime X470 Pro and AMD Ryzen 7 5800X3D
And yes, it's probably this older combination. I resigned and sent back the A380. Perhaps in future with a completely new platform.

**Reply by kayti** [reply]
> @kayti I assume the limiting factor was the CPU or the motherboard since PCIE4 8x is theoretically capable of ~15 GB/s [Wikipedia] and 4k@120Hz HDR would need around 32.27 Gbps [[trychen.com](https://trychen.com/feature/video-bandwidth)]. Assuming raw video data needs to be transferred to (from the render GPU to the encoding GPU) and from (to the network interface for transmission) the encoding GPU PCIE x8 can't be the limiting factor here.
> 
> What's the model of your CPU/Motherboard that you used for testing?

I need to reassess my old answer :) Probably, I had some compatibility problems with the previous Motherboard (some GIGABYTE, don't remember model now). 

Recently I started permanently using Dual dGPU setup. CPU Ryzen 7900, Motherboard ROG Strix B650E-F, 32 GB DDR5 6000. 
Motherboard has 16x and 4x PCIe4.
RTX 3080 in 16x slot, A380 in 4x slot.
My stream is 3440x1440@120 HDR usually.
Using latest apollo, latest drivers. A380 used for encoding (fastest setting, AV1).
RTX 3080 used for games.

Moonlight set at 300 Mbps. 
It works well enough in terms of latency compared to a pure RTX 3080 as an encoder. I didn't feel any noticeable additional lag.

HAGS works well with this setup, RTX 3080 has some more memory to spare for game, framerate is higher and more stable.

But 3440@1440@165 HDR was already probably pushing A380 encoding capabilities, I had some occasional latency spikes. I run some benchmarks for encoding and average encoding framerate was ~200FPS but with dips around 145. Probably this is where I felt those spikes.

**Reply by kayti** [reply]
@tibia1426 Check my newer anwser: https://github.com/ClassicOldSong/Apollo/discussions/155#discussioncomment-13573097

**Reply by mreilaender** [reply]
Love to hear it works since I'm currently evaluating a similar setup :) What client do you use for decoding?

**Reply by kayti** [reply]
For 3440x1440 usually miniPC with 7840HS. Practically zero problems, sub 1ms decoding AV1 streams, but the don't have hardware 4:4:4 decoder.
Another client is 1345U Laptop, similar like client before, decodes AV1 like a champ. Has 4:4:4 decoder, but in Apollo it works only with HEVC, so I don't use it anyway as I default to AV1.
N100 is far weaker, but able to handle 4k60 well, as I have it connected to 4k60 projector. Never tried it with higher resolution/refresh.

**Reply by mreilaender** [reply]
What problems did u run into when using AV1 encoding with Apollo? Was it hardware or software related?

**Reply by kayti** [reply]
Right now I don't have any problems. Before probably I had some compatibility problem with motherboard. I changed platform and everything worked just fine without any specific tweaks (other that choosing encoding GPU in apollo and setting rtx as high performance card in windows).

**Reply by TheEldritchVoid** [reply]
I'm currently rocking an Rtx 2060 and an rx580 dual dGPU as a new perm update to my setup (adding in my old card since there's a few new tools that can take advantage of it) from what I'm seeing here if I run into hiccups it's probably on the mobo side at this point?
